internal/procedures/server/start.go

```diff
@@ -47,19 +45,69 @@ func Start(initiator common.ServerInitiator) (err error) {
 		go startCourse(course)
 	}
 
-	// Cleanup any temp dirs.
-	defer util.RemoveRecordedTempDirs()
+	return nil
+}
 
-	err = server.RunServer(initiator)
-	if err != nil {
-		return fmt.Errorf("Error during server startup sequence: '%w'.", err)
+func CleanupAndStop() (err error) {
+	common.Lock(SERVER_LOCK)
+	defer common.Unlock(SERVER_LOCK)
+
+	if apiServer == nil {
+		return nil
 	}
 
+	err = errors.Join(err, db.Close())
+	err = errors.Join(err, util.RemoveRecordedTempDirs())
+
+	apiServer.Stop()
+
+	apiServer = nil
+
 	log.Debug("Server closed.")
 
 	return err
 }
 
+func assignAndSetupServer(initiator common.ServerInitiator, skipSetup bool) error {
+	common.Lock(SERVER_LOCK)
+	defer common.Unlock(SERVER_LOCK)
+
+	apiServer = server.NewAPIServer()
+
+	if !skipSetup {
+		err := setup(initiator)
+		if err != nil {
+			return fmt.Errorf("Failed to setup the server: '%w'.", err)
+		}
+	}
+
+	return nil
+}
+
+func RunAndBlock(initiator common.ServerInitiator, skipSetup bool) (err error) {
```

Comment:
Our typical pattern is to hide advanced options like skipSetup() using a *Full() function.